### PR TITLE
perf: optimize qwen-vl with symm mem allreduce

### PR DESCRIPTION
**Pull request by yuan-luo**
## Motivation

The All-Reduce in VL model inference has a high proportion.
image

Qwen2.5-VL-72B uses inplace all-reduce, by default is NCCL AR which is inefficient.

image

This PR introduces in-place Symm Mem All-Reduce for Qwen2.5-VL model. (Literally for any other models using in-place AR) Improved 2% e2e input throughput. Single AR latency drops from 446.9us to 209.1us.
For Qwen3-VL model, improved 3.9% e2e input throughput.

```
python -m sglang.launch_server \
  --model-path /home/admin/Qwen2.5-VL-7B-Instruct \
  --host 0.0.0.0 --port 30000 --trust-remote-code \
  --tp-size 4 --enable-cache-report  --log-level info \
  --max-running-requests 64 --mem-fraction-static 0.65 \
  --chunked-prefill-size 8192  --attention-backend fa3 \
  --chat-template qwen2-vl --mm-attention-backend fa3 
  --enable-torch-symm-mem
```
Main:
image

PR:
image

```
python -m python.sglang.bench_serving \
    --backend sglang \
    --dataset-name random-image \
    --random-output-len 1 \
    --num-prompts 768 --max-concurrency 16 --random-image-resolution 360p --profile 

MAIN:
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 16        
Successful requests:                     768       
Benchmark duration (s):                  69.40     
Total input tokens:                      422368    
Total generated tokens:                  718       
Total generated tokens (retokenized):    677       
Request throughput (req/s):              11.07     
Input token throughput (tok/s):          6085.89   
Output token throughput (tok/s):         10.35     
Total token throughput (tok/s):          6096.24   
Concurrency:                             15.89     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   1435.81   
Median E2E Latency (ms):                 1426.29   
---------------Time to First Token----------------
Mean TTFT (ms):                          1268.69   
Median TTFT (ms):                        1397.04   
P99 TTFT (ms):                           1804.01   
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================

PR:
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 16        
Successful requests:                     768       
Benchmark duration (s):                  68.52     
Total input tokens:                      422846    
Total generated tokens:                  718       
Total generated tokens (retokenized):    660       
Request throughput (req/s):              11.21     
Input token throughput (tok/s):          6171.31   
Output token throughput (tok/s):         10.48     
Total token throughput (tok/s):          6181.79   
Concurrency:                             15.87     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   1415.47   
Median E2E Latency (ms):                 1434.92   
---------------Time to First Token----------------
Mean TTFT (ms):                          1220.77   
Median TTFT (ms):                        1390.59   
P99 TTFT (ms):                           1759.99   
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================


```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Review by mickqian — COMMENTED** [review]
Great! Could you apply to qwen3-vl as well?

File: python/sglang/srt/managers/schedule_batch.py [hunk 1]
```diff
@@ -1762,7 +1762,11 @@ def filter_batch(
         self.seq_lens_cpu = self.seq_lens_cpu[keep_indices]
         self.orig_seq_lens = self.orig_seq_lens[keep_indices_device]
         self.out_cache_loc = None
-        self.seq_lens_sum = self.seq_lens.sum().item()
+        if isinstance(self.seq_lens_cpu, torch.Tensor):
```
Comment (BBuf) [inline python/sglang/srt/managers/schedule_batch.py:1765]: When will `self.seq_lens_cpu` is a numpy array?

File: python/sglang/srt/layers/rotary_embedding.py [hunk 2]
```diff
@@ -1020,12 +1031,14 @@ def __init__(
...
         dtype: torch.dtype,
         mrope_section: Optional[List[int]] = None,
+        mrope_interleaved: bool = False,
     ) -> None:
         super().__init__(
             head_size, rotary_dim, max_position_embeddings, base, is_neox_style, dtype
         )
 
         self.mrope_section = mrope_section
+        self.mrope_interleaved = mrope_interleaved
```
Comment (BBuf) [inline python/sglang/srt/layers/rotary_embedding.py:1041]: Can we add a unit-test for `mrope_interleaved` ?

Comment (yuan-luo) [reply python/sglang/srt/layers/rotary_embedding.py:1041]: mrope_interleaved is a dependency for my next PR mrope kernel. I will add a unit test there. In this PR it is just a place-holder.

File: python/sglang/srt/managers/schedule_batch.py (hunk 1 above)
Comment (yuan-luo) [reply python/sglang/srt/managers/schedule_batch.py:1765]: Here is a safe-guard. self.seq_lens_cpu has been reassigned in many places:
```
self.seq_lens_cpu = torch.cat([self.seq_lens_cpu, other.seq_lens_cpu])
....
self.seq_lens_cpu = torch.tensor(seq_lens, dtype=torch.int64)
...
self.seq_lens_cpu = torch.empty(0, dtype=torch.int64)
...
self.seq_lens_cpu = self.seq_lens_cpu[keep_indices]
```
The handling of the numpy array here is a kind of `last resort or fallback` logic without any harm.

File: python/sglang/srt/layers/rotary_embedding.py (hunk 2 above)
Comment (BBuf) [reply python/sglang/srt/layers/rotary_embedding.py:1041]: Ok

File: python/sglang/srt/managers/schedule_batch.py (hunk 1 above)
Comment (BBuf) [reply python/sglang/srt/managers/schedule_batch.py:1765]: Make sense

**Review by BBuf — APPROVED** [review]

**Review by mickqian — APPROVED** [review]

**Comment by yuan-luo** [comment]
Qwen3-VL-30B-A3B-Instruct obtains better performance:
tp8, bs16, input 360p image, output 1
Input throughput 3.9% speedup (29112 -> 30317)
Request throughput 4.1% speedup (52.9->55.1)

Baseline:
```
$python -m sglang.launch_server --model-path /home/admin/Qwen3-VL-30B-A3B-Instruct --host 0.0.0.0 --port 30000 --trust-remote-code --tp-size 8 --enable-cache-report --log-level info --max-running-requests 64 --mem-fraction-static 0.65 --chunked-prefill-size 8192 --attention-backend fa3 --chat-template qwen2-vl --mm-attention-backend fa3

$python -m python.sglang.bench_serving \
>     --backend sglang \
>     --dataset-name random-image \
>     --random-output-len 1 \
>     --num-prompts 768 --max-concurrency 16 --random-image-resolution 360p

============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 16        
Successful requests:                     768       
Benchmark duration (s):                  14.51     
Total input tokens:                      422485    
Total generated tokens:                  715       
Total generated tokens (retokenized):    642       
Request throughput (req/s):              52.92     
Input token throughput (tok/s):          29112.57  
Output token throughput (tok/s):         49.27     
Total token throughput (tok/s):          29161.84  
Concurrency:                             15.76     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   297.79    
Median E2E Latency (ms):                 256.81    
---------------Time to First Token----------------
Mean TTFT (ms):                          241.52    
Median TTFT (ms):                        244.79    
P99 TTFT (ms):                           1888.75   
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================
```

PR:
```
$python -m sglang.launch_server --model-path /home/admin/Qwen3-VL-30B-A3B-Instruct --host 0.0.0.0 --port 30000 --trust-remote-code --tp-size 8 --enable-cache-report --log-level info --max-running-requests 64 --mem-fraction-static 0.65 --chunked-prefill-size 8192 --attention-backend fa3 --chat-template qwen2-vl --mm-attention-backend fa3 --enable-torch-symm-mem

$python -m python.sglang.bench_serving \
>     --backend sglang \
>     --dataset-name random-image \
>     --random-output-len 1 \
>     --num-prompts 768 --max-concurrency 16 --random-image-resolution 360p


============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 16        
Successful requests:                     768       
Benchmark duration (s):                  13.93     
Total input tokens:                      422441    
Total generated tokens:                  709       
Total generated tokens (retokenized):    643       
Request throughput (req/s):              55.12     
Input token throughput (tok/s):          30317.07  
Output token throughput (tok/s):         50.88     
Total token throughput (tok/s):          30367.96  
Concurrency:                             15.77     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   286.18    
Median E2E Latency (ms):                 240.81    
---------------Time to First Token----------------
Mean TTFT (ms):                          234.52    
Median TTFT (ms):                        227.50    
P99 TTFT (ms):                           1996.16   
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================
```

**Comment by yuan-luo** [comment]
gsm8k acc result:
```
Baseline:
$python3 benchmark/gsm8k/bench_sglang.py --num-questions 200 --parallel 128 --num-shots 8 --port 30000 --data-path ./test.jsonl
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:14<00:00, 13.55it/s]
Accuracy: 0.950
Invalid: 0.000
Latency: 14.891 s
Output throughput: 1955.172 token/s

PR:
$python3 benchmark/gsm8k/bench_sglang.py --num-questions 200 --parallel 128 --num-shots 8 --port 30000 --data-path ./test.jsonl
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:14<00:00, 13.90it/s]
Accuracy: 0.950
Invalid: 0.000
Latency: 14.485 s
Output throughput: 2008.234 token/s
```

**Comment by ch-wan** [comment]
This PR breaks the vllm dependency test.

**Comment by yuan-luo** [comment]
When I ran the CI, the vllm dependency test has been pending to wait for a running task for about one hour. So I canceled this test case. Not realize it happens to break this test. I'll fix it and submit again.
